### PR TITLE
Partially fixes #197

### LINKED ISSUE #197: NMS.KK.OBD.000 The outbound-dialer module shall create a daily file of outgoing calls
4.3.1 TargetFile Notification API
NMS shall invoke this API to notify IVR platform when a target file is ready.

### PR DESCRIPTION
**Pull request by frankhuster**
Lots of work to be done still. But smaller chunks would be best.

### CODE REVIEW

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallRetry.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package org.motechproject.nms.outbounddialer.domain;
+
+import org.motechproject.mds.annotations.Entity;
+import org.motechproject.mds.annotations.Field;
+
+@Entity(tableName = "nms_call_retry")
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallRetry.java]: nam_call_retries? :)

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallStage.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package org.motechproject.nms.outbounddialer.domain;
+
+public enum CallStage {
+    Fresh,
+    Retry1,
+    Retry2,
+    Retrry3,
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallStage.java]: typo. Retrrrrrrrrrrrrrrryyyyyyyyyyyyyyyy

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallRetry.java (hunk 1 above)
Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallRetry.java]: Err, yep.

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallStage.java (hunk 2 above)
Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/CallStage.java]: Gr

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+public class TargetFileAudit {
+    @Field
+    private String fileName;
+
+    @Field
+    private Integer recordCount;
+
+    @Field
+    private String checksum;
+
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java]: Do you want to have a field in here to record the CDRs received at the end of the day as well? That way, we can easily correlate the data? probably feature creep for this checkin though

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java]: I'm thinking I'll add a CdrFileAudit entity actually

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/CdrFileService.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ *
+ */
+public interface CdrFileService {
+
+    /**
+     * The controller's cdrFileNotification http endpoint was invoked by the IVR system, start processing the provided
+     * CDR files
+     *
+     * @param request
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/CdrFileService.java]: nitpick: request explanation?

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/TargetFileNotification.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package org.motechproject.nms.outbounddialer.service;
+
+/**
+ * Request body
+ *
+ * 4.3.1 TargetFile Notification API
+ * NMS shall invoke this API to notify IVR platform when a target file is ready.
+ * http://<IVROBDAPI:port>/obdmanager/notifytargetfile
+ */
+public class TargetFileNotification {
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/TargetFileNotification.java]: Service suffix to the class name?

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/CdrFileService.java (hunk 4 above)
Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/CdrFileService.java]: I haven't dealt with CDRs yet. Lots left to do.

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/TargetFileNotification.java (hunk 5 above)
Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/TargetFileNotification.java]: This is not s service class, just a DTO used to send in the POST body to the IVR system

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        String intervalProp = settingsFacade.getProperty(TARGET_FILE_MS_INTERVAL);
+        Long msInterval = Long.parseLong(intervalProp);
+
+        LOGGER.debug(String.format("The %s message will be sent every %sms starting %s",
+                GENERATE_TARGET_FILE_EVENT, msInterval.toString(), today.toString()));
+
+        //Schedule repeating job
+        MotechEvent event = new MotechEvent(GENERATE_TARGET_FILE_EVENT);
+        RepeatingSchedulableJob job = new RepeatingSchedulableJob(
+                event,          //MOTECH event
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: Not fully sure about the inner workings of schedulable job, but just to confirm, this has no dependency on -

1. the time that the service is first deployed
 - so today vs. timeProp?
2. the running time of the event once its fired
- it doesnt keep sliding.

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        this.subscriptionDataService = subscriptionDataService;
+        this.callRetryDataService = callRetryDataService;
+
+        scheduleTargetFileGeneration();
+    }
+
+
+    private String targetFileName() {
+        DateTime today = DateTime.now();
+        return String.format("OBD_%04d%02d%02d%02d%02d%02d.csv",
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: nitpick. set this format as a constant or a property?

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java (hunk 6 above)
Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: 0.
  - Clearly needs more comments.

1.
  - today: today's day
  - timProp: time at which the event will be fired today.

2.
  - should the process take more than 24h we'd be in trouble, yep...

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java (hunk 7 above)
Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: yep, will do

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+                        new QueryParams(page, maxQueryBlock));
+                numBlockRecord = subscriptions.size();
+
+                for (Subscription subscription : subscriptions) {
+                    writer.write(subscription.getSubscriptionId());
+                    writer.write(",");
+                    writer.write(subscription.getSubscriber().getCallingNumber().toString());
+                    writer.write(",");
+                    writer.write(subscription.getLanguage().getCode());
+                    //todo...
```
Comment (llavoie) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: I know this isn't done yet -- but just a reminder to include the subscriptionMode (MCTS or IVR) in the file, as mentioned in #180

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: Yep, will do next checkin when I merge & have access to that property on the Subscription object.

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+    @MotechListener(subjects = { GENERATE_TARGET_FILE_EVENT })
+    public void generateTargetFile(MotechEvent event) {
+        LOGGER.info(event.toString());
+
+        TargetFileNotification tfn = generateTargetFile();
+
+        if (tfn != null) {
+            //notify the IVR system the file is ready
+            sendNotificationRequest(tfn);
+        }
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: when tfn == null, we raise alerts and let this fall through, correct? Is there a backup/catchup mechanism or are the calls for the day given up?

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: yep, the idea is to raise alarms and do nothing...
'caus there's nothing we can really do.

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+    }
+
+
+    private void sendNotificationRequest(TargetFileNotification tfn) {
+        String notificationUrl = settingsFacade.getProperty(TARGET_FILE_NOTIFICATION_URL);
+        LOGGER.info("Sending {} to {}", tfn, notificationUrl);
+
+        try {
+            HttpClient httpClient = HttpClients.createDefault();
+            HttpPost httpPost = new HttpPost(notificationUrl);
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: Just curious, have you considered using the SimpleHttpClient here?

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: SimpleHttpClient is just used for testing. It doesn't do what we want here.

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        Long msInterval = Long.parseLong(intervalProp);
+
+        LOGGER.debug(String.format("The %s message will be sent every %sms starting %s",
+                GENERATE_TARGET_FILE_EVENT, msInterval.toString(), today.toString()));
+
+        //Schedule repeating job
+        MotechEvent event = new MotechEvent(GENERATE_TARGET_FILE_EVENT);
+        RepeatingSchedulableJob job = new RepeatingSchedulableJob(
+                event,          //MOTECH event
+                today.toDate(), //startTime
```
Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: ok, to confirm yet again :)

the schedulable event is enqueued when the service initializes. The job is passed a parameter in with today.ToDate() which has a direct dependency on the time that the service is deployed, right?

Wouldn't you want to give it target file time?

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: Nope.
Look at line 72, I set the `today` variable to today's date but set its hour and minute to what I got from the settings file.

Comment (koshalt) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: ah right. sneaky

tricked

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/service/impl/TargetFileServiceImpl.java]: But I did add a comment. For the likes of you.

**Comment by koshalt** [comment]
lgtm. push up changes and merge?

File: outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
+package org.motechproject.nms.outbounddialer.domain;
+
+import org.motechproject.mds.annotations.Entity;
+import org.motechproject.mds.annotations.Field;
+
+@Entity(tableName = "nms_target_file_audit")
```
Comment (larubbio) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java]: I'm late to the game, but table names should be plural

Comment (frankhuster) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java]: What's the plural for audit?

Comment (larubbio) [inline outbound-dialer/src/main/java/org/motechproject/nms/outbounddialer/domain/TargetFileAudit.java]: audits

https://en.wiktionary.org/wiki/audit

**Pull request merged**